keps/sig-api-machinery/00xx-publish-crd-openapi.md

```diff
@@ -0,0 +1,142 @@
+documentation for CRs.
+
+## Motivation
+
+Publishing CRD OpenAPI enables client-side validation, schema explanation and
+client generation for CRs. It covers the gap between CRD and native Kubernetes
+APIs, which already support OpenAPI documentation.
+
+Publishing CRD OpenAPI is also noted as potential followup in [Validation for CustomResources Implementation Plan](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/customresources-validation.md#implementation-plan).
+
+### Goals
+
+* For every CRD that we serve, publish Paths (operations that we support on
+  resource and subresources) and Definitions (for both CR object and CR list
+  object) in OpenAPI documentation to fully demonstrate the existence of the
+  API.
+* For CRDs with validation JSON-Schema defined, the CR object Definition should
+  include both CRD validation JSON-Schema and native Kubernetes ObjectMeta and
+  TypeMeta properties.
+* For CRDs without validation JSON-Schema, the CR object Definition will be an
+  emtpy object which validate arbitrary request on client-side.
+
+### Non-Goals
+
+* Support/honor OpenAPI AdditionalProperties in kube-openapi and kubectl, to
+  enable client-side validation on native Kubernetes ObjectMeta and TypeMeta
+  while allowing arbitrary extra fields in a request.
+* Finalize the design on how we aggregate OpenAPI documentation from
+  apiextensions-apiserver
+  - [Short term design doc](https://docs.google.com/document/d/13lBj8Stdwku8BgL0fbT__4Iw97NRh77loJ_MoZuCGwQ/edit#)
+    including multiple workarounds has been purposed and is under discussion.
+  - Long term design doc including generalizing APIService API or implementing
+    watch mechanism for /openapi/v2 endpoint on apiextension-apiserver is work
+    in progress.
+
+## Proposal
+
+ref: https://github.com/kubernetes/kubernetes/pull/71192
+
+### Risks and Mitigations
+
+TODO(roycaihw):
+
+if somehow the OpenAPI documentation for native APIs gets removed, or the code aggregates a huge
+number/size of OpenAPI documentation because of a bug? kube-aggregator might
+(not likely with sttts' optimization) get overwhelmed? kubectl (and anything else that downloads
+openapi from a cluster) might (likely) get overwhelmed?
+
+are we under potential risk of leaking information from the CRD object?
+
+TODO(roycaihw): also do we need to test the behavior above?
+
+mitigate:
+
+currently we use the same feature flag as we use for the json-schema:
+CustomResourceValidation. the feature flag can be turned off to stop
+openapi creation & aggregation & publishing from happening. should we use
+another feature flag separately?
+
+## Graduation Criteria
+
+In 1.14, publishing CRD OpenAPI should be e2e-tested to ensure:
+
+* for CRD with validation schema
+
+1. `kubectl create` client-side validation works to:
+  - allow requests with known and required properites
+  - reject requests with unknown properties
```

Comment:
clarify "reject requests with unknown properties when disallowed by the schema"

Reply:
changed

correct me if I am wrong, it seems that we need to support (at least) additionalProperties to conform this statement, but additionalProperties is out of scope for this KEP